File: Spark_Recommendation_content/task2_1.scala
import org.apache.spark.sql.SparkSession
import scala.math._

object task2_1 {

  def computePearson(ratingsB: Map[String, Double], ratingsI: Map[String, Double], minCommon: Int = 1): Double = {
    val commonUsers = ratingsB.keySet.intersect(ratingsI.keySet)
    if (commonUsers.size < minCommon) return 0.0

    val avgB = commonUsers.map(ratingsB).sum / commonUsers.size
    val avgI = commonUsers.map(ratingsI).sum / commonUsers.size

    val numerator = commonUsers.map(u => (ratingsB(u) - avgB) * (ratingsI(u) - avgI)).sum
    val denominatorB = sqrt(commonUsers.map(u => pow(ratingsB(u) - avgB, 2)).sum)
    val denominatorI = sqrt(commonUsers.map(u => pow(ratingsI(u) - avgI, 2)).sum)
    val denominator = denominatorB * denominatorI

    if (denominator == 0.0) {
      if (numerator == 0.0) 1.0 else -1.0
    } else {
      math.min(1.0, numerator / denominator)
    }
  }

  def predictRating(user: String, business: String,
                    businessRatings: Map[String, Map[String, Double]],
                    userRatings: Map[String, Map[String, Double]],
                    businessMeans: Map[String, Double],
                    userMeans: Map[String, Double],
                    overallAvg: Double,
                    neighborLimit: Int = 50): Double = {

    if (!businessRatings.contains(business)) {
      return userRatings.get(user).map(_ => userMeans.getOrElse(user, overallAvg)).getOrElse(overallAvg)
    }
    if (!userRatings.contains(user)) {
      return businessMeans.getOrElse(business, overallAvg)
    }

    val neighborList = userRatings(user).toList.flatMap { case (b_i, rating) =>
      if (b_i == business || !businessRatings.contains(b_i)) None else {
        val sim = computePearson(businessRatings(business), businessRatings(b_i)) * (if (computePearson(businessRatings(business), businessRatings(b_i)) < 0) 0.8 else 1.2)
        val normalizedRating = rating - (businessMeans.getOrElse(b_i, overallAvg) + userMeans.getOrElse(user, overallAvg) - overallAvg)
        Some((sim, normalizedRating))
      }
    }.sortBy(-_._1).take(neighborLimit)

    val numerator = neighborList.map { case (sim, rating) => sim * rating }.sum
    val denominator = neighborList.map { case (sim, _) => abs(sim) }.sum

    val baseline = overallAvg +
      (businessMeans.getOrElse(business, overallAvg) - overallAvg) +
      (userMeans.getOrElse(user, overallAvg) - overallAvg)

    val prediction = if (denominator != 0) baseline + numerator / denominator else baseline
    math.max(1.0, math.min(5.0, prediction))
  }

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder.appName("task2_1").getOrCreate()
    val sc = spark.sparkContext

    val trainFile = args(0)
    val testFile = args(1)
    val outputFile = args(2)

    val rdd = sc.textFile(trainFile)
    val header = rdd.first()

    val train = rdd.filter(_ != header).map(_.split(",")).cache()

    val businessRatings = train.map(x => (x(1), (x(0), x(2).toDouble)))
      .groupByKey()
      .mapValues(_.toMap)

    val businessMeans = businessRatings.mapValues(v => v.values.sum / v.size).collect().toMap

    val userRatings = train.map(x => (x(0), (x(1), x(2).toDouble)))
      .groupByKey()
      .mapValues(_.toMap)

    val userMeans = userRatings.mapValues(v => v.values.sum / v.size).collect().toMap

    val overallAvg = train.map(_(2).toDouble).mean()

    // 使用广播变量
    val broadcastBusinessRatings = sc.broadcast(businessRatings.collect().toMap)
    val broadcastUserRatings = sc.broadcast(userRatings.collect().toMap)
    val broadcastBusinessMeans = sc.broadcast(businessMeans)
    val broadcastUserMeans = sc.broadcast(userMeans)
    val broadcastOverallAvg = sc.broadcast(overallAvg)

    val testData = sc.textFile(testFile)
      .filter(_ != header)
      .map(_.split(",")).map(x => (x(0), x(1)))

    val output = testData.map { case (user, business) =>
      val rating = predictRating(user, business, broadcastBusinessRatings.value, broadcastUserRatings.value, broadcastBusinessMeans.value, broadcastUserMeans.value, broadcastOverallAvg.value)
      (user, business, rating)
    }.collect()

    val writer = new java.io.PrintWriter(outputFile)
    writer.println("user_id,business_id,prediction")
    output.foreach { case (u, b, p) =>
      writer.println(s"$u,$b,$p")
    }
    writer.close()

    train.unpersist() // 释放缓存资源
    spark.stop()
  }
}

File: Spark_Recommendation_content/task2_1.py
from pyspark.sql import SparkSession
import json
import sys
import math


def compute_pearson(ratings_b, ratings_i, min_common=1):
    common_users = set(ratings_b.keys()) & set(ratings_i.keys())

    # not consider when there is only on customer
    if len(common_users) < min_common:
        return 0.0

    avg_b = sum(ratings_b[u] for u in common_users) / len(common_users)
    avg_i = sum(ratings_i[u] for u in common_users) / len(common_users)

    numerator = sum((ratings_b[u] - avg_b) * (ratings_i[u] - avg_i) for u in common_users)
    denominator_b = math.sqrt(sum((ratings_b[u] - avg_b) ** 2 for u in common_users))
    denominator_i = math.sqrt(sum((ratings_i[u] - avg_i) ** 2 for u in common_users))
    denominator = denominator_b * denominator_i

    if denominator == 0:
        if numerator == 0:
            return 1.0
        else:
            return -1.0
    return min(1.0, numerator / denominator)


def predict_rating_for_pair(user, business, business_sets, user_sets, business_means, user_means, overall_avg, neighbor_limit=15):

    # if business_id not in the test set

    if business not in business_sets:
        if user in user_sets:
            return user_means.get(user, overall_avg)
        else:
            return overall_avg
    # if user_id not in the test set
    if user not in user_sets:
        return business_means.get(business, overall_avg)

    neighbor_list = []
    user_ratings = user_sets[user]
    for b_i, rating in user_ratings.items():
        if b_i == business:
            continue
        if b_i in business_sets:
            sim = compute_pearson(business_sets[business], business_sets[b_i])
            if sim < 0:
                sim = sim * 0.8
            else:
                sim = sim * 1.2

            normalized_rating = rating - (business_means.get(b_i, overall_avg) + user_means.get(user, overall_avg) - overall_avg)
            neighbor_list.append((sim, normalized_rating))

    neighbor_list = sorted(neighbor_list, key=lambda x: x[0], reverse=True)[:neighbor_limit]

    numerator = sum(sim * rating for sim, rating in neighbor_list)
    denominator = sum(abs(sim) for sim, rating in neighbor_list)

    baseline = overall_avg + (business_means.get(business, overall_avg) - overall_avg) + (user_means.get(user, overall_avg) - overall_avg)

    if denominator != 0:
        prediction = baseline + numerator / denominator
    else:
        prediction = baseline

    return min(max(prediction, 1.0), 5.0)



def main(train_file, test_file, output_file):
    spark = SparkSession.builder.appName("task2_1").getOrCreate()
    sc = spark.sparkContext

    rdd = sc.textFile(train_file)
    header = rdd.first()

    business_rdd = rdd.filter(lambda x: x != header).map(lambda x: x.split(',')).map(lambda x: (x[1], (x[0], float(x[2]))))
    business_sets = business_rdd.groupByKey().mapValues(lambda vals: {u: r for u, r in vals}).collectAsMap()

    business_means = {b: sum(ratings.values()) / len(ratings) for b, ratings in business_sets.items()}

    user_rdd = rdd.filter(lambda x: x != header).map(lambda x: x.split(',')).map(lambda x: (x[0], (x[1], float(x[2]))))
    user_sets = user_rdd.groupByKey().mapValues(lambda vals: {b: r for b, r in vals}).collectAsMap()

    overall_avg = rdd.filter(lambda x: x != header).map(lambda x: float(x.split(',')[2])).mean()

    user_means = {u: sum(ratings.values()) / len(ratings) for u, ratings in user_sets.items()}
    
    
    broadcast_user_means = sc.broadcast(user_means)
    broadcast_business = sc.broadcast(business_sets)
    broadcast_user = sc.broadcast(user_sets)
    broadcast_business_means = sc.broadcast(business_means)
    broadcast_overall = sc.broadcast(overall_avg)

    test_rdd = sc.textFile(test_file)
    test_header = test_rdd.first()

    test_data = test_rdd.filter(lambda x: x != test_header).map(lambda x: x.split(',')).map(lambda x: (x[0], x[1]))

    predictions = test_data.map(lambda pair: (
        pair[0],
        pair[1],
        predict_rating_for_pair(
            pair[0],
            pair[1],
            broadcast_business.value,
            broadcast_user.value,
            broadcast_business_means.value,
            broadcast_user_means.value,  
            broadcast_overall.value,
            neighbor_limit=50
        )
    ))


    output = predictions.collect()
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("user_id,business_id,prediction\n")
        for (b1, b2, sim) in output:
            f.write(f"{b1},{b2},{sim}\n")

    spark.stop()


if __name__ == "__main__":
    train_file = sys.argv[1]
    test_file = sys.argv[2]
    output_file = sys.argv[3]
    main(train_file, test_file, output_file)
